internal/storage/image.go

```diff
@@ -420,6 +438,209 @@ func (svc *imageService) imageStatus(systemContext *types.SystemContext, unstabl
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return "", fmt.Errorf("create policy context: %w", err)
+	}
+
+	defer func() {
+		if err := policyContext.Destroy(); err != nil {
+			log.Errorf(ctx, "Unable to destroy policy: %v", err)
+		}
+	}()
+
+	localManifestDigest, err := svc.pullLocalManifestDigest(ctx, systemContext, policyContext, imageParsed)
+	if err != nil {
+		return "", fmt.Errorf("pull local manifest digest: %w", err)
+	}
+	log.Debugf(ctx, "Local manifest digest for image %q is: %s", imageParsed, localManifestDigest)
+
+	canonicalRef, err := reference.WithDigest(reference.TrimNamed(imageParsed), localManifestDigest)
+	if err != nil {
+		return "", fmt.Errorf("create canonical reference: %w", err)
+	}
+
+	return canonicalRef.String(), nil
+}
+
+func (svc *imageService) pullLocalManifestDigest(ctx context.Context, sys *types.SystemContext, policyContext *signature.PolicyContext, imageParsed reference.Named) (digest.Digest, error) {
+	srcRef, err := docker.NewReference(imageParsed)
+	if err != nil {
+		return "", fmt.Errorf("creating docker:// reference for %q: %w", imageParsed.String(), err)
+	}
+
+	destRef, err := istorage.Transport.NewStoreReference(svc.store, imageParsed, "")
+	if err != nil {
+		return "", fmt.Errorf("creating containers-storage: reference for %q: %w", imageParsed.String(), err)
+	}
+
+	manifestBytes, err := copy.Image(ctx, policyContext, destRef, srcRef, &copy.Options{
+		SourceCtx:      sys,
+		DestinationCtx: sys,
+	})
+	if err != nil {
+		var e errcode.Error
+		if as := errors.As(err, &e); as && e.ErrorCode().Descriptor().HTTPStatusCode == http.StatusUnauthorized {
+			// We cannot support private registries right now, but may with future support from the CRI.
+			log.Errorf(ctx, "Verifying policies against private registries it not supported yet: %v", err)
+			return "", nil
+		}
+
+		var policyErr signature.PolicyRequirementError
+		if errors.As(err, &policyErr) {
+			return "", fmt.Errorf("%w: %w", crierrors.ErrSignatureValidationFailed, err)
+		}
+
+		return "", fmt.Errorf("pulling %q to %q: %w", transports.ImageName(srcRef), transports.ImageName(destRef), err)
+	}
+	return digest.FromBytes(manifestBytes), nil
+}
+
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string) error {
+	policy, err := signature.DefaultPolicy(systemContext)
+	if err != nil {
+		return fmt.Errorf("get default policy: %w", err)
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+
+	defer func() {
+		if err := policyContext.Destroy(); err != nil {
+			log.Errorf(ctx, "Unable to destroy policy: %v", err)
+		}
+	}()
+
+	var localManifestDigest digest.Digest
+	// On fresh pulls retrieve the full image and digest as configImage
+	if parsedConfigImage, err := reference.Parse(configImage); err == nil {
```

Comment:
(Push parsing heuristics up the call stack. This is too deep to be ignoring parsing errors. I suspect this will have to change anyway to resolve the races in `container_create.go`.